### PR TITLE
Add element selector to template-part block

### LINKED ISSUE #27092: Allow selecting HTML element for the template-part wrapper
Pages need elements like `<header>`, `<main>`, `<footer>` etc. Right now that's not possible from the UI, so the result is HTML that is not semantic and in some cases that can also be an a11y issue.

**Describe the solution you'd like**
Users should be able to select the "type" of their template-parts. The available options should be:
* `<header>`
* `<main>`
* `<section>`
* `<article>`
* `<aside>`
* `<footer>`

It could be done on group elements etc, but that would be a lot more confusing for the end user. Having that option in the template-part block would allow some degree of freedom without overcomplicating things. It will also promote some good practices, since it will enforce using template-parts for separate sections.

Without that option, the generated HTML is a collection of `<div>` elements which is bad both for accessibility and SEO.

### PR DESCRIPTION
**Pull request by aristath**
## Description
Fixes #27092 

This PR adds a `SelectControl` to the `template-part` block, to allow selecting what the template-part's wrapper element should be.

## How has this been tested?
Tested in an FSE theme in the site editor:
* Selected the theme's template parts, confirmed that I was able to change them to their proper semantically-correct values
* Saved & reloaded the editor, confirmed that values persist
* Removed the patch and confirmed nothing breaks (backwards-compatible)
* Confirmed that the selected values are properly mirrored on the frontend.

## Screenshots 
Peek 2020-11-19 15-42

## Types of changes
* Added a `SelectControl` with values that are applicable to template-parts

Most of the changes in the diff are indentation changes because adding the control required nesting the elements inside a wrapper.

There are no breaking changes. We already had a `tagName` property in the block, but it required manual intervention in the HTML - which was impossible to do since the site-editor doesn't have a code view.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by gziolo** [comment]
For a better context, we tried something similar with the Group block that has `tagName` attributes. We explored making it possible to pick different tag names through block variations exposed in the inserter but we didn't proceed with the solution eventually, reverted code for reference:

https://github.com/WordPress/gutenberg/pull/20218/commits/de6b2ed21843fce5729486080b38d139eb5b37af

**Comment by carolinan** [comment]
So far I have used block variations for template-parts as a way to speed up the development when I test FSE themes and for me it has worked really well. 

Despite this, I think the drop down is a better solution:

- It is easy to change the element.
- The variations can be confusing for non developers, especially if a template part with the variation name "header" can be confused with selectable template parts for site headers.

I suggest the following changes:
Place the semantics under the Advanced section to avoid confusion for users who are not familiar with HTML.

Remove the word "Wrapping" and only keep "HTML element"

File: packages/block-library/src/template-part/edit/index.js [hunk 1]
```diff
@@ -23,7 +29,7 @@ import TemplatePartPlaceholder from './placeholder';
 import TemplatePartSelection from './selection';
 
 export default function TemplatePartEdit( {
-	attributes: { postId: _postId, slug, theme, tagName: TagName = 'div' },
+	attributes: { postId: _postId, slug, theme, tagName: TagName = 'section' },
```
Comment (carolinan) [inline packages/block-library/src/template-part/edit/index.js:32]: If the default is div, shouldn't this remain div, not section?

**Review by gziolo — APPROVED** [review]
Code wise, it's good to go after addressing the nitpick about the default value in the `TemplatePartEdit` component prop.

Moving forward, we will need a strategy to introduce unified handling for blocks to act as wrapping elements:
- Group
- Post Content 
- Query
- Reusable Block
- Template Part

or blocks that can have different wrapping tag elements depending on function:
- Heading
- List

It would be great to hear from @mtias or @jasmussen on the above before we commit to anything.

(The failing e2e test was improved after the last commit pushed from this branch, rebase should resolve it)

**Comment by jasmussen** [comment]
Instead of adding a new panel called "Semantics", can we add this to the "Advanced" section?

File: packages/block-library/src/template-part/edit/index.js (hunk 1 above)
Comment (aristath) [reply packages/block-library/src/template-part/edit/index.js]: Changed default to `div` :+1:

**Comment by aristath** [comment]
> Instead of adding a new panel called "Semantics", can we add this to the "Advanced" section?

Done in https://github.com/WordPress/gutenberg/pull/27101/commits/5f2fa20c85c5ba86cf7f09b620bca344242bb083 :+1:

**Comment by aristath** [comment]
> Remove the word "Wrapping" and only keep "HTML element"

Done in https://github.com/WordPress/gutenberg/pull/27101/commits/1b0c37decd36fad49d458de86a893a2412ab38dc :+1:

**Comment by jasmussen** [comment]
I think there's an opportunity to further refine this with some of the component work and systems that are outlined in #27331, but so long as this is in the advanced section and is collapsed by default, seems fine to move forward with this one as a feature addition. Nice work.

**Comment by gziolo** [comment]
I have just tested it and it work according to the latest suggestion from @jasmussen. Let's get this in 🎉

**Comment by aristath** [comment]
Thank you @gziolo for your testing, I'll go ahead and merge it :+1:

**Pull request merged**